### PR TITLE
Add support for requiring reason for access requests

### PR DESCRIPTION
**Pull request by kopiczko**
Issue https://github.com/gravitational/teleport/issues/20164

TODO:
- [ ] https://github.com/gravitational/teleport/pull/49363
- [x] https://github.com/gravitational/teleport/pull/49256
- [x] https://github.com/gravitational/teleport/pull/49258

changelog: Added support for requiring reason for access requests (with a new `role.spec.allow.request.reason.mode` setting).

### CODE REVIEW

File: lib/services/access_request_test.go [hunk 1]
```diff
@@ -2142,6 +2143,256 @@ func TestAutoRequest(t *testing.T) {
...
+			expectError: nil,
+		},
+		{
+			name:         "resource request: but require reason when another role allowing _role_ access requires reason for the role",
+			currentRoles: []string{"fork-node-requester", "fork-access-requester-with-reason"},
+			requestResourceIDs: []types.ResourceID{
+				{ClusterName: clusterName, Kind: types.KindNode, Name: "fork-node"},
+			},
+			expectError: trace.BadParameter(`request reason must be specified (required for role "fork-access")`),
+		},
```
Comment (kopiczko) [inline lib/services/access_request_test.go:2300]: This is the situation discussed here https://gravitational.slack.com/archives/C07QJQNGY2J/p1731620401319179

**Review by smallinsky — COMMENTED** [review]
First pass. 

Could you take a look at [UT failing](https://github.com/gravitational/teleport/actions/runs/11899332569/job/33158729460?pr=49124) on the AccessRequest test cases; 

```
=== FAIL: lib/auth TestAccessRequestNonGreedyAnnotations/identity-resource-requester_requests_resource,_receives_identity_annotations (0.03s)
{"caller":"services/access_checker.go:1269","component":null,"level":"warning","message":"Failed to find roles in x509 identity for test-user. Fetching from backend. If the identity provider allows username changes, this can potentially allow an attacker to change the role of the existing user.","timestamp":"2024-11-18T19:32:40Z"}
{"timestamp":"2024-11-18T19:32:40Z","level":"debug","caller":"events/discard.go:147","message":"Discarding event","event_id":"","event_type":"role.created","event_time":"0001-01-01T00:00:00Z","event_index":0}
{"caller":"auth/auth.go:5216","component":"auth","level":"debug","message":"Creating Access Request 019340c3-4a76-7f2a-bf1e-70abfac19a85 with expiry 2024-11-18 20:32:39 +0000 UTC.","timestamp":"2024-11-18T19:32:40Z"}
    auth_with_roles_test.go:8628: 
        	Error Trace:	/__w/teleport/teleport/lib/auth/auth_with_roles_test.go:8628
        	Error:      	Received unexpected error:
        	            	no roles configured in the "search_as_roles" for this user allow access to any requested resources. The user may already have access to all requested resources with their existing roles. resources: ["/localhost/node/server-identity"] roles: [identity-access] login: ""
        	Test:       	TestAccessRequestNonGreedyAnnotations/identity-resource-requester_requests_resource,_receives_identity_annotations
      ```

**Comment by kopiczko** [comment]
Sorry for the pings I did a accidental master merge again 🙈

File: lib/auth/auth_with_roles_test.go
Comment (kopiczko) [inline lib/auth/auth_with_roles_test.go:1]: Those changes are unrelated. It just happens that my changes exposed those broken tests. 

Unfortunately, it took me whole day to figure out why they were broken. I was trying to blame wrong things at first.

**Comment by kopiczko** [comment]
@smallinsky unit tests are fixed with https://github.com/gravitational/teleport/pull/49124/commits/8fb0891f2c919e82ac42705a087d950c85b0520e. They were broken and those changes exposed that, but unfortunately it took me whole day to fix it.

There are still some docs linter problems I'll have to fight with and the rest of the review remarks I'll address tomorrow.

**Review by smallinsky — COMMENTED** [review]
@kopiczko 

Screenshot 2024-11-20 at 08 56 17

Could you split this PR into 2-3 separate PRs?

 - (optional) Proto Changes
 - Actual Implementation 
 - Doc Alignment

**Comment by kopiczko** [comment]
Ok, will split. Moving it to draft for now.

**Comment by kopiczko** [comment]
I'll probably have to combine Proto Changes and Doc Alignment. Otherwise linter will be sad https://github.com/gravitational/teleport/actions/runs/11922379049/job/33228509916

**Comment by kopiczko** [comment]
First 2 extracted PRs:

- Add request.reason.required to role conditions (proto only) [#49256](https://github.com/gravitational/teleport/pull/49256)
- Fix broken auth Access Request creation tests [#49258](https://github.com/gravitational/teleport/pull/49258)

The remaining big here is the implementation and (related) tests.

File: lib/services/access_request.go [hunk 2]
```diff
@@ -275,7 +275,14 @@ func CalculateAccessCapabilities(ctx context.Context, clock clockwork.Clock, clt
 		caps.SuggestedReviewers = v.SuggestedReviewers
 	}
 
-	caps.RequireReason = v.requireReason
+	// If empty-string reason is valid, then reason is not required.
+	err = v.validateReason(ctx, "", caps.RequestableRoles, req.ResourceIDs)
+	if trace.IsBadParameter(err) {
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply lib/services/access_request.go]: I changed the method to accommodate for this https://github.com/gravitational/teleport/pull/49124/commits/564cca7329443ad257a552720b0805ab3ef9dad1. I'm not 100% happy with the `explanation string` return value, but I'd rather have it printed for the user than having a prettier method signature.

File: lib/services/access_request.go [hunk 3]
```diff
@@ -1561,17 +1609,26 @@ func setAllowRequestKubeResourceLookup(allowKubernetesResources []types.RequestK
 // push compiles a role's configuration into the request validator.
 // All of the requesting user's statically assigned roles must be pushed
 // before validation begins.
-func (m *RequestValidator) push(ctx context.Context, role types.Role) error {
+func (m *RequestValidator) push(ctx context.Context, userState UserState, role types.Role) error {
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply lib/services/access_request.go]: https://github.com/gravitational/teleport/pull/49124/commits/564cca7329443ad257a552720b0805ab3ef9dad1

File: lib/services/access_request.go [hunk 4]
```diff
@@ -269,13 +262,31 @@ func CalculateAccessCapabilities(ctx context.Context, clock clockwork.Clock, clt
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
+	} else {
+		caps.ApplicableRolesForResources, err = v.applicableSearchAsRoles(ctx, req.ResourceIDs, req.Login)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
 	}
```
Comment (kopiczko) [inline lib/services/access_request.go:270]: Correct me if I'm wrong, but it's either `RequestableRoles` if the flag is set or `ApplicableRolesForResources` if no flag is set. The original condition was confusing.

**Comment by kopiczko** [comment]
@smallinsky @r0mant @fspmarshall I extracted and merged the dependencies, adapted to request.mode change and consider this ready to review now.

File: lib/services/access_request.go (hunk 4 above)
Comment (kopiczko) [reply lib/services/access_request.go]: I was wrong! Had to revert that. Luckily it's covered by integration tests.

File: lib/services/access_request.go [hunk 5]
```diff
@@ -275,7 +275,20 @@ func CalculateAccessCapabilities(ctx context.Context, clock clockwork.Clock, clt
...
+		caps.RequireReason, _, err = v.isReasonRequired(ctx, caps.RequestableRoles, nil)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	} else {
+		caps.RequireReason, _, err = v.isReasonRequired(ctx, caps.ApplicableRolesForResources, nil)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
```
Comment (smallinsky) [inline lib/services/access_request.go:290]: Can we extract this logic into a separate function, something like: 
```suggestion
func (v ...) calcRequireReasonCap(ctx context.Context, req types.AccessCapabilitiesRequest, caps types.AccessCapabilities) (bool, error) {
	if v.requireReasonForAllRoles {
		return true, nil
	}
	if req.RequestableRoles {
		return v.isReasonRequired(ctx, caps.RequestableRoles, nil)
	}
	return v.isReasonRequired(ctx, caps.ApplicableRolesForResources, nil)
}
```

**Review by smallinsky — APPROVED** [review]

File: lib/services/access_request.go [hunk 6]
```diff
@@ -1132,8 +1169,14 @@ func (m *RequestValidator) Validate(ctx context.Context, req types.AccessRequest
 		return trace.BadParameter("request validator configured for different user (this is a bug)")
 	}
 
-	if m.requireReason && req.GetRequestReason() == "" {
-		return trace.BadParameter("request reason must be specified (required by static role configuration)")
+	if len(req.GetRequestReason()) == 0 {
```
Comment (r0mant) [inline lib/services/access_request.go:1172]: I think we should do strings.TrimSpace on this to prevent basic cheating like a bunch of whitespaces.

**Review by r0mant — APPROVED** [review]

File: lib/services/access_request.go [hunk 7]
```diff
@@ -1564,14 +1630,23 @@ func setAllowRequestKubeResourceLookup(allowKubernetesResources []types.RequestK
...
 	}
 
 	allow, deny := role.GetAccessRequestConditions(types.Allow), role.GetAccessRequestConditions(types.Deny)
 
+	if allow.Reason != nil && allow.Reason.Mode.Required() {
+		for _, r := range allow.Roles {
+			m.requiringReasonRoles[r] = struct{}{}
+		}
+		for _, r := range allow.SearchAsRoles {
+			m.requiringReasonRoles[r] = struct{}{}
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply lib/services/access_request.go:1651]: No, that doesn't hurt. The value will be `struct{}{}` regardless.

File: lib/services/access_request.go [hunk 8]
```diff
@@ -1313,6 +1349,36 @@ func (m *RequestValidator) Validate(ctx context.Context, req types.AccessRequest
...
+		// Do not provide loginHint. We want all matching search_as_roles for those resources.
+		roles, err := m.applicableSearchAsRoles(ctx, requestedResourceIDs, "")
+		if err != nil {
+			return false, "", trace.Wrap(err)
+		}
+		if len(allApplicableRoles) == 0 {
+			allApplicableRoles = roles
+		} else {
+			allApplicableRoles = append(allApplicableRoles, roles...)
+		}
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply lib/services/access_request.go:1375]: Yes, I wanted to avoid allocation, but it doesn't matter much.

File: lib/services/access_request.go (hunk 6 above)
Comment (kopiczko) [reply lib/services/access_request.go]: https://github.com/gravitational/teleport/pull/49124/commits/b9389c45d7820013389eba731191bb1ea99dc6ba

File: lib/services/access_request.go [hunk 9]
```diff
@@ -1313,6 +1326,41 @@ func (m *RequestValidator) Validate(ctx context.Context, req types.AccessRequest
 	return nil
 }
 
+// validateReason checks if the reason is valid and returns BadParameterError if it is not. It may
+// return any other error.
+func (m *RequestValidator) validateReason(ctx context.Context, reason string, requestedRoles []string, requestedResourceIDs []types.ResourceID) error {
+	// Non-empty reason is always valid.
+	if len(reason) > 0 {
```
[the remark this answers is not in the document]
Comment (kopiczko) [reply lib/services/access_request.go]: https://github.com/gravitational/teleport/pull/49124#discussion_r1858496766
